jdbc/src/main/java/com/interface21/jdbc/core/JdbcTemplate.java

```diff
@@ -14,4 +19,58 @@ public class JdbcTemplate {
     public JdbcTemplate(final DataSource dataSource) {
         this.dataSource = dataSource;
     }
+
+    public <T> T queryForObject(String sql, RowMapper<T> mapper, Object... args) {
+        try (Connection conn = dataSource.getConnection();
+             PreparedStatement pstmt = conn.prepareStatement(sql)) {
+            setPreparedStatement(pstmt, args);
+            ResultSet rs = pstmt.executeQuery();
+
+            if (rs.next()) {
+                return mapper.mapRow(rs, 1);
+            }
+            return null;
+
+        } catch (SQLException e) {
+            throw new RuntimeException("DB 조회에 실패했습니다. :" + sql, e);
+        }
+    }
+
+    public <T> List<T> query(String sql, RowMapper<T> mapper) {
+        try (Connection conn = dataSource.getConnection();
+             PreparedStatement pstmt = conn.prepareStatement(sql)) {
+            final ResultSet rs = pstmt.executeQuery();
+
+            final List<T> result = new ArrayList<>();
+            int rowNum = 1;
+            while (rs.next()) {
+                result.add(mapper.mapRow(rs, rowNum++));
+            }
+            return result;
+        } catch (SQLException e) {
+            throw new RuntimeException("DB 조회에 실패했습니다. :" + sql, e);
+        }
+    }
+
+    public void update(String sql, Object... args) {
+        try (Connection conn = dataSource.getConnection();
+             PreparedStatement pstmt = conn.prepareStatement(sql)) {
+            setPreparedStatement(pstmt, args);
+            pstmt.executeUpdate();
+        } catch (SQLException e) {
+            throw new RuntimeException("DB 조회에 실패했습니다. :" + sql, e);
+        }
+    }
+
+    private void setPreparedStatement(PreparedStatement pstmt, Object[] args) {
+        int paramCount = 0;
+        try {
+            paramCount = pstmt.getParameterMetaData().getParameterCount();
+            for (int i = 0; i < args.length; i++) {
+                pstmt.setObject(i + 1, args[i]);
+            }
+        } catch (SQLException ex) {
+            throw new IllegalArgumentException("SQL 파라미터 개수 불일치: expected " + paramCount + ", actual " + args.length);
+        }
```

Comment:
파라미터를 순회하기 전에 개수를 먼저 명시적으로 검증하면 좋을것 같습니다. 왜냐하면 catch 모든 SQL 예외를 잡고 있기 때문에 파라미터 개수 불일치 예외가 타입 불일치, null 제약 조건 위반 등의 다른 예외와 구분이 안될 것 같아요 ! 그래서 다음과 같이 수정될 것 같은데 어떠신가여??

```suggestion
      try {
          int paramCount = pstmt.getParameterMetaData().getParameterCount();
          if (paramCount != args.length) {
              throw new IllegalArgumentException(
                  "SQL 파라미터 개수 불일치: expected " + paramCount + ", actual " + args.length
              );
          }

          for (int i = 0; i < args.length; i++) {
              pstmt.setObject(i + 1, args[i]);
          }
      } catch (SQLException ex) {
          throw new RuntimeException("SQL 파라미터 설정 실패: " + ex.getMessage(), ex);
      }
```

Reply:
헉 코드 예시까지 넘 감사합니다!
그렇네용 파라미터가 안맞는 경우는 자주 발생할 것 같아서 예외를 꼭 던지고 싶었는데 exception 범위를 생각하지 못했네요!
반영했습니닷